src/Momento.Sdk/Internal/Middleware/RetryMiddleware.cs

```diff
@@ -0,0 +1,89 @@
+﻿using System;
+using System.Threading.Tasks;
+using Grpc.Core;
+using Microsoft.Extensions.Logging;
+using Momento.Sdk.Config.Middleware;
+using System.Linq;
+using Momento.Protos.CacheClient;
+using System.Collections.Generic;
+
+namespace Momento.Sdk.Config.Retry
+{
+    internal class RetryMiddleware : IMiddleware
+    {
+        public ILoggerFactory? LoggerFactory { get; }
+
+        private readonly ILogger _logger;
+        private readonly IRetryStrategy _retryStrategy;
+
+        public RetryMiddleware(ILoggerFactory loggerFactory, IRetryStrategy retryStrategy)
+        {
+            LoggerFactory = loggerFactory;
+            _logger = loggerFactory.CreateLogger<RetryMiddleware>();
+            _retryStrategy = retryStrategy;
+        }
+
+        public RetryMiddleware WithLoggerFactory(ILoggerFactory loggerFactory)
+        {
+            return new(loggerFactory, _retryStrategy);
+        }
+
+        IMiddleware IMiddleware.WithLoggerFactory(ILoggerFactory loggerFactory)
+        {
+            return WithLoggerFactory(loggerFactory);
+        }
+
+        public async Task<MiddlewareResponseState<TResponse>> WrapRequest<TRequest, TResponse>(
+            TRequest request,
+            CallOptions callOptions,
+            Func<TRequest, CallOptions, Task<MiddlewareResponseState<TResponse>>> continuation
+        ) where TRequest : class where TResponse : class
+        {
+            var foo = request.GetType();
+            MiddlewareResponseState<TResponse> nextState;
+            int attemptNumber = 0;
+            int? retryAfterMillis = 0;
+            do
+            {
+                var delay = retryAfterMillis ?? 0;
+                if (delay > 0)
+                {
+                    await Task.Delay(delay);
+                }
+                attemptNumber++;
+                nextState = await continuation(request, callOptions);
```

Comment:
I'm not sure I see how this will disambiguate between the likes of ListPushBack() or DictionaryIncrement() (non-idempotent methods) and Get() or Set() (idempotent methods). What's your plan there?

Reply:
See the code in `DefaultRetryEligibilityStrategy`, which is called as part of the `FixedCountRetryStrategy`.

Reply:
I completely missed `_retryableRequestTypes`, thanks.